tests/unittests/cmd/devel/test_logs.py

```diff
@@ -35,6 +35,7 @@ def test_collect_logs_with_userdata_requires_root_user(
             " Try sudo cloud-init collect-logs\n" == m_stderr.getvalue()
         )
 
+    @pytest.mark.allow_all_subp
```

Comment:
Allowing subp to run make the unit tests dependent on the underlying system. Instead, I think we should mock the call to `Init`.

Reply:
I am not quite sure what's the easiest and best way to implement a mock for `Init` but i took a look at unittests implementation for "clean.py" and this looks very promising to me:
`@pytest.fixture(scope="function")
def init_class(clean_paths):
    class FakeInit:
        cfg = {
            "def_log_file": clean_paths.log,
            "output": {"all": f"|tee -a {clean_paths.output_log}"},
        }
        # Ensure cloud_dir has a trailing slash, to match real behaviour
        paths = MyPaths(cloud_dir=f"{clean_paths.cloud_dir}/")

        def __init__(self, ds_deps):
            pass

        def read_cfg(self):
            pass

    return FakeInit`
Am i far off track or is this something we could work with? thank you for your help

Reply:
Here is a patch mocking `Init` and `get_log_files` to test the isolated behavior of `collect-logs`.  The patch also includes a unit test covering the new behavior of `get_log_files`.

```diff
diff --git a/tests/unittests/cmd/devel/test_logs.py b/tests/unittests/cmd/devel/test_logs.py
index b7ea33052..004b4957f 100644
--- a/tests/unittests/cmd/devel/test_logs.py
+++ b/tests/unittests/cmd/devel/test_logs.py
@@ -35,7 +35,6 @@ class TestCollectLogs:
             " Try sudo cloud-init collect-logs\n" == m_stderr.getvalue()
         )
 
-    @pytest.mark.allow_all_subp
     def test_collect_logs_creates_tarfile(self, m_getuid, mocker, tmpdir):
         """collect-logs creates a tarfile with all related cloud-init info."""
         m_getuid.return_value = 100
@@ -57,6 +56,12 @@ class TestCollectLogs:
         )
         output_tarfile = str(tmpdir.join("logs.tgz"))
 
+        mocker.patch(M_PATH + "Init", autospec=True)
+        mocker.patch(
+            M_PATH + "get_config_logfiles",
+            return_value=[log1, log1_rotated, log2, log2_rotated],
+        )
+
         date = datetime.utcnow().date().strftime("%Y-%m-%d")
         date_logdir = "cloud-init-logs-{0}".format(date)
 
@@ -147,7 +152,6 @@ class TestCollectLogs:
         )
         fake_stderr.write.assert_any_call("Wrote %s\n" % output_tarfile)
 
-    @pytest.mark.allow_all_subp
     def test_collect_logs_includes_optional_userdata(
         self, m_getuid, mocker, tmpdir
     ):
@@ -167,6 +171,12 @@ class TestCollectLogs:
         )
         output_tarfile = str(tmpdir.join("logs.tgz"))
 
+        mocker.patch(M_PATH + "Init", autospec=True)
+        mocker.patch(
+            M_PATH + "get_config_logfiles",
+            return_value=[log1, log2],
+        )
+
         date = datetime.utcnow().date().strftime("%Y-%m-%d")
         date_logdir = "cloud-init-logs-{0}".format(date)
 
diff --git a/tests/unittests/test_util.py b/tests/unittests/test_util.py
index 7186780ba..9e7d06378 100644
--- a/tests/unittests/test_util.py
+++ b/tests/unittests/test_util.py
@@ -2344,6 +2344,30 @@ class TestGetConfigLogfiles(helpers.CiTestCase):
             ),
         )
 
+    def test_output_logs_parsed_when_teeing_files_and_rotated(self):
+        """When output configuration is parsed when teeing files and rotated
+        log files are present."""
+        tmpd = self.tmp_dir()
+        log1 = self.tmp_path("my.log", tmpd)
+        log1_rotated = self.tmp_path("my.log.1.gz", tmpd)
+        log2 = self.tmp_path("himom.log", tmpd)
+        log2_rotated = self.tmp_path("himom.log.1.gz", tmpd)
+
+        util.write_file(log1_rotated, "hello")
+        util.write_file(log2_rotated, "hello")
+
+        self.assertEqual(
+            [log2, log2_rotated, log1, log1_rotated],
+            sorted(
+                util.get_config_logfiles(
+                    {
+                        "def_log_file": str(log1),
+                        "output": {"all": f"|tee -a {log2}"},
+                    }
+                )
+            ),
+        )
+
 
 class TestMultiLog(helpers.FilesystemMockingTestCase):
     def _createConsole(self, root):
```

Reply:
thank you for your help :) . i applied your patch. Furthermore, i enhanced integration tests.